### ISSUE TITLE
Sometimes RA reports errors that `rustc` doesn't agree exist

### ISSUE THREAD

**Issue by Lokathor** [rust-lang/rust-analyzer#18542]
Lately, in various places in my project I get errors like

> parsing.rs
> expected 2 arguments, found 1 [Ln 176, Col 6]

but the thing is: there's *not* a wrong number of arguments. the project builds with `cargo build` and runs with `cargo run` and even shows correct output. So why the heck is RA telling me about fake errors. I thought it was something to do with like, bad cached data or something, but even after deleting the target/ directory the same problem happened a few days later.

- **rust-analyzer version**: rust-analyzer version: 0.3.2188-standalone (ba56d9b9b 2024-11-17) [c:\Users\Daniel\.vscode\extensions\rust-lang.rust-analyzer-0.3.2188-win32-x64\server\rust-analyzer.exe]
- **rustc version**: rustc 1.82.0-nightly (6de928dce 2024-08-18)
- **editor or extension**: VSCode `rust-analyzer` extension
- **relevant settings**: ??? I've no clue what could be related here.
- **repository link (if public, optional)**: https://github.com/Lokathor/yagbas/tree/1e2e182548b144c1d47de03126d3fc49d8f9d89b

**Comment by lnicola** [comment]
Can you post a screenshot of the error? r-a has its own diagnostics, which sometimes have false positives. It happens with `Into`, for example.

**Comment by Lokathor** [comment]
![Image](https://github.com/user-attachments/assets/a8a0ef3d-7808-4b92-bb51-776f115e8655)

**Comment by lnicola** [comment]
Yeah, given the bounds on that function, I'm not really surprised. Adding explicit type annotations usually works around this.

**Comment by Lokathor** [comment]
uh, where? to the closure?

**Comment by lnicola** [comment]
Maybe, or to what `fn_body` returns (moving it to a variable).

**Comment by Lokathor** [comment]
the type of `fn_body` is...

> let fn_body: Labelled<NestedIn<Collect<SeparatedBy<MapWith<impl Parser<'_, I, Statement, Full<Rich<'_, TokenTree, FileSpan>, (), ()>> + Clone, Statement, impl Fn(Statement, &mut MapExtra<'_, '_, I, Full<Rich<'_, TokenTree, FileSpan>, (), ()>>) -> FileSpanned<Statement>>, Repeated<impl Parser<'_, I, (), Full<Rich<'_, TokenTree, FileSpan>, (), ()>> + Clone, (), I, Full<Rich<'_, TokenTree, FileSpan>, (), ()>>, FileSpanned<Statement>, (), I, Full<Rich<'_, TokenTree, FileSpan>, (), ()>>, FileSpanned<Statement>, {unknown}>, impl Parser<'_, I, I, Full<Rich<'_, TokenTree, FileSpan>, (), ()>> + Clone, I, Full<Rich<'_, TokenTree, FileSpan>, (), ()>, {unknown}, Full<Rich<'_, TokenTree, FileSpan>, (), ()>>, &str>

**Comment by lnicola** [comment]
Try Go to definitions on those methods in the chain, one will probably be from the wrong trait.

**Comment by Lokathor** [comment]
I didn't write the `chumsky` library i'm using here, but all the steps in the chain seem to go to the right traits.

**Comment by lnicola** [comment]
Ouch, we're so slow on that file. The problem is on `map`, we pick the wrong one:

![Image](https://github.com/user-attachments/assets/88b1816d-e4e0-49d6-a59b-34c1acdbe10f)

Explicitly choosing the trait fixes the false positive:

```
  let x = Parser::map(
    kw_fn_p().ignore_then(name).then(args).then(fn_body),
    |((name, arguments), statements)| Function { name, arguments, statements },
  )
  .labelled("function")
  .as_context();
```

**Comment by ChayimFriedman2** [comment]
Closing since this is likely no longer applicable. @Lokathor if this is still reproduced please reopen.
